plutus-tx-plugin/src/Language/PlutusTx/Plugin.hs

```diff
@@ -63,7 +65,45 @@ data PluginOptions = PluginOptions {
     }
 
 plugin :: GHC.Plugin
-plugin = GHC.defaultPlugin { GHC.installCoreToDos = install, GHC.pluginRecompile = GHC.flagRecompile }
+plugin = GHC.defaultPlugin { GHC.pluginRecompile = GHC.flagRecompile
+                           , GHC.installCoreToDos = install
+                           }
+    where
+      install :: [GHC.CommandLineOption] -> [GHC.CoreToDo] -> GHC.CoreM [GHC.CoreToDo]
+      install args rest = do
+          -- create simplifier pass to be placed at the front
+          simplPass <- mkSimplPass <$> GHC.getDynFlags
+          -- instantiate our plugin pass
+          pluginPass <- mkPluginPass
+                       <$> parsePluginArgs args
+                       <*> getMarkerName
+          -- return the pipeline
+          pure $
+             simplPass
+             : pluginPass
+             : rest
+
+      -- <https://hackage.haskell.org/package/ghc-8.10.1/docs/GhcPlugins.html#v:thNameToGhcName> says
+      -- that we only have to call this once, since it is bound to an exact th symbol (using the 'quote syntax)
```

Comment:
I *think* this is okay, but it's subtle. `thNameToGhcName` can look up the name in the current `HscEnv`, which differs *per module*. So we might find the name or not depending whether we're in the context of a particular module when this is run. 

I don't know when `install` is run, but I'm suspicious. Whereas the previous version would only call it once we were definitely processing the module in question. So either we need a careful justification of this, or go back to the previous version which was more obviously correct (and that should anyway have had a note explaining this).

Reply:
I see your point now, and I understand this is not the optimal behaviour, since we may traverse looking for markers in a module for no reason, since 'plc is not in the local-module context. According to ghc doc: 

> thNameToGhcName :: Name -> CoreM (Maybe Name)Source#
Attempt to convert a Template Haskell name to one that GHC can understand. Original TH names such as those you get when you use the 'foo syntax will be translated to their equivalent GHC name exactly. Qualified or unqualified TH names will be dynamically bound to names in the module being compiled, if possible. Exact TH names will be bound to the name they represent, exactly.

The documentation is a bit confusing actually. I have a feeling that even if we move this `getMarkerName 'plc` to its older place, we do get the same suboptimal behaviour according to the docs. Since this is an "original th name" with `'foo` syntax, it means that even if we move it to be executed at the local-module level, failing to find the 'plc in the local context means that it will try in the global context as well. We perhaps have to manually construct a "Qualified or unqualified TH names" that  "will be **dynamically bound** to names in the module being compiled". Then if this manual th name symbol is not found locally, we do not have to traverse the whole binds looking for its occurences inside expressions.
This is all speculation  from just reading the ghc docs. I did not dig that deep in the ghc source code.

Reply:
I feel that using the 'foo syntax results to "exact GHC names", whereas perhaps using the (mkName "Module.plc") syntax results to dynamically bound names. But then dynamically bound names might have problems of their own.

Reply:
I moved it back at the location it used to be. I added a TODO: note about it.